Improve value of INVALID_PFN

krytarowski opened the pull request:
pfn is a 40bit field and a strict compiler complains about implicit
narrowing of value into cointainer. Change the constant to represent
the greatest 40-bit integer.

No functional change intended.

AlexAltea commented:
> pfn is a 40bit field

Just curious, what's the source for this? I haven't seen any references to "Page Frame Number" or "PFN" in any of the Intel SDM volumes.

> strict compiler complains about implicit narrowing of value into cointainer

Has this happened at all? I thought we were using `uint64_t`'s for PFN values.

krytarowski commented:
```
$ git grep pfn|grep 40
core/include/ept2.h:        uint64 pfn           : 40;  // bits 51..12
core/include/ept2.h:        uint64 pfn          : 40;  // bits 51..12
```

```
#   compile  netbsd/ept_tree.o
gcc -O2 -g   -std=gnu99   -Werror   -ffreestanding  -fno-strict-aliasing -Wno-pointer-sign -mno-red-zone -mno-mmx -mno-sse -mno-avx -msoft-float -mcmodel=kernel -fno-omit-frame-pointer    -nostdinc -I. -I/public/haxm/netbsd -isystem /usr/src/sys -isystem /usr/src/sys/arch -isystem /usr/src/sys/../common/include -D_KERNEL -D_LKM -D_MODULE -DSYSCTL_INCLUDE_DESCR -c    ../core/ept_tree.c
In file included from ../core/ept_tree.c:35:0:
../core/include/paging.h:80:23: error: large integer implicitly truncated to unsigned type [-Werror=overflow]
 #define INVALID_ADDR  ~(uint64)0
                       ^
../core/include/paging.h:81:21: note: in expansion of macro ‘INVALID_ADDR’
 #define INVALID_PFN INVALID_ADDR
                     ^~~~~~~~~~~~
../core/ept_tree.c:38:12: note: in expansion of macro ‘INVALID_PFN’
     .pfn = INVALID_PFN
            ^~~~~~~~~~~
cc1: all warnings being treated as errors
*** Error code 1

Stop.
make: stopped in /public/haxm/netbsd
```

AlexAltea commented:
Nice catch! (Also, I didn't know about git-grep, that's pretty useful)

raphaelning commented on core/include/paging.h:
```diff
@@ -76,7 +76,7 @@ static paddr_t get_pagebase(vaddr_t p, uint order)
 
 #define INVALID_ADDR  ~(uint64_t)0
-#define INVALID_PFN   INVALID_ADDR
+#define INVALID_PFN   ((uint64_t)(~0ULL) >> 24) /* 40-bit field */
```
nit: Any good reason to use `(uint64_t)` and `ULL` at the same time? If there is, please also update `INVALID_ADDR` above. Actually, `INVALID_PFN` can be derived from `INVALID_ADDR`:

```suggestion
#define INVALID_PFN   (INVALID_ADDR >> 24)  /* 40-bit field */
```

krytarowski replied:
For amd64 / i386 ABIs this ULL with uint64_t isn't required. I will follow the suggestion.

raphaelning commented:
BTW, I just realized that the commit message is missing the `Signed-off-by` line. We require it:

https://github.com/intel/haxm/blob/master/CONTRIBUTING.md

krytarowski commented:
Done.

AlexAltea commented:
It seems the your branch `krytarowski:netbsd-5` contains commits already merged into `intel:master`, namely 5717abb, c332eec, ecf0f22. Maybe it should be rebased properly?

krytarowski commented:
According to my local copy, everything is fine.

```
commit 35afef5e49ecc63793e883bd26c10ff5007d5704 (HEAD -> netbsd-5)
Author: Kamil Rytarowski <n54@gmx.com>
Date:   Sat Nov 17 18:49:17 2018 +0100

    Improve value of INVALID_PFN
    
    pfn is a 40bit field and a strict compiler complains about implicit
    narrowing of value into cointainer. Change the constant to represent
    the greatest 40-bit integer.
    
    No functional change intended.
    
    Signed-off-by: Kamil Rytarowski <n54@gmx.com>

commit eda0250588dbd2ba9e427b1b9154d082e7f904f4 (upstream/master, origin/master, origin/HEAD, master)
Merge: 5e8b93c f4884d4
Author: Yu Ning <raphael.ning@gmail.com>
Date:   Mon Nov 19 18:03:44 2018 +0800

    Merge pull request #124 from kryptoslogic/bt-tests
    
    Added tests for BT/BTS/BTR/BTC instructions
```

If there is something accounted extra, please cherry-pick the proper commit.

raphaelning commented:
@krytarowski Sorry I just edited out the email addresses in the git log from your comment. I hope you don't mind.

It's indeed weird that GitHub shows the already-merged commits while saying "This branch has no conflicts with the base branch." I don't think it allows us to cherry-pick one commit. But we'll try to merge this PR after testing and see what happens.

AlexAltea commented:
@krytarowski I get different results (see text below). Try force-pushing it?
```
$ git clone --single-branch -b netbsd-5 https://github.com/krytarowski/haxm
$ cd haxm/
$ git status
On branch netbsd-5
Your branch is up to date with 'origin/netbsd-5'.

nothing to commit, working tree clean
$ git log
commit 6e1bb29eb1dc8e5c7ab352a511554554c3783f44 (HEAD -> netbsd-5, origin/netbsd-5)
Author: Kamil Rytarowski <n54@gmx.com>
Date:   Sat Nov 17 18:49:17 2018 +0100

    Improve value of INVALID_PFN
    
    pfn is a 40bit field and a strict compiler complains about implicit
    narrowing of value into cointainer. Change the constant to represent
    the greatest 40-bit integer.
    
    No functional change intended.
    
    Signed-off-by: Kamil Rytarowski <n54@gmx.com>

commit ecf0f2297eb360f88f280de0722f00a723cbe48f
Author: Alexandro Sanchez Bach <asanchez@kryptoslogic.com>
Date:   Sat Nov 17 12:27:29 2018 +0100

    Added tests for BT/BTS/BTR/BTC instructions
    
    Signed-off-by: Alexandro Sanchez Bach <asanchez@kryptoslogic.com>
```

krytarowski commented:
I've recreated branch from scratch on top of master, cherry-picked the relevant commit, removed old branch, renamed current to the pushed branch to this review, and push forced. It looks like it's fine now.
